mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala

```diff
@@ -112,12 +114,15 @@ final class CategoricalSplit private[ml] (
     }
   }
 
-  override def equals(o: Any): Boolean = {
-    o match {
-      case other: CategoricalSplit => featureIndex == other.featureIndex &&
-        isLeft == other.isLeft && categories == other.categories
-      case _ => false
-    }
+  override def hashCode(): Int = {
+    val state = Seq(featureIndex, isLeft, categories)
+    state.map(Objects.hashCode).foldLeft(0)((a, b) => 36 * a + b)
```

Comment:
Nit: isn't something like 31 used as the multiplier because it's prime? while any prime is probably OK maybe 36 is a tiny bit less optimal